Feature/lp 114 repair request viewmodel

JoseGR98 opened the pull request:
LP-114   Repair request - ViewModel
https://devaptivist.atlassian.net/browse/LP-114?atlOrigin=eyJpIjoiMTlmZDc0MzlhMGUyNDE2NjhlNTA3YmUyN2ZjMjc3Y2QiLCJwIjoiaiJ9

ViewModel worked for the RepairView. Several changes:
Classes:
RepairModel --> rename to RepairRequest
UIEventsRepair --> to handle events type

Navigation:
ApartmentGraph --> small changes

RepairUI:
RepairViewModel --> big changes, added events to handle data manipulation
RepairView --> small changes, adapted to Events

Commit:
Merge branch 'develop' into feature/lp-114-repair-request-viewmodel

Commit:
Progress on basic structure of ViewModel

Commit:
Merge branch 'develop' into feature/lp-114-repair-request-viewmodel

Commit:
Merge branch 'develop' into feature/lp-114-repair-request-viewmodel

Commit:
Merge branch 'develop' into feature/lp-114-repair-request-viewmodel

Commit:
Merge branch 'develop' into feature/lp-114-repair-request-viewmodel

Commit:
Merge branch 'develop' into feature/lp-114-repair-request-viewmodel

Commit:
Merge branch 'develop' into feature/lp-114-repair-request-viewmodel

Commit:
Merge branch 'develop' into feature/lp-114-repair-request-viewmodel

# Conflicts:
#	feature/repair/src/main/java/com/iteneum/repair/data/RepairViewModel.kt

Commit:
Merge branch 'develop' into feature/lp-114-repair-request-viewmodel

Commit:
Merge branch 'develop' into feature/lp-114-repair-request-viewmodel

Commit:
Merge branch 'develop' into feature/lp-114-repair-request-viewmodel

Commit:
LP-114 ModelView first progress

Commit:
LP-114 ModelView progress

Commit:
Merge branch 'develop' into feature/lp-114-repair-request-viewmodel

Commit:
Merge branch 'develop' into feature/lp-114-repair-request-viewmodel

Commit:
RepairViewModel.kt progress

Commit:
Merge branch 'develop' into feature/lp-114-repair-request-viewmodel

JoseGR98 committed:
Merge branch 'develop' into feature/lp-114-repair-request-viewmodel

JoseGR98 committed:
RepairModel.kt added.
RepairViewModel.kt almost complete (verify field unused)
RepairView adapted to ViewModel.
ApartmentNavGraph.kt modified to work with RepairView.

JoseGR98 committed:
Build.gradle of Repair Updated

JoseGR98 committed:
TODO added

JoseGR98 committed:
Merge branch 'develop' into feature/lp-114-repair-request-viewmodel

JoseGR98 committed:
RepairModel reworked
RepairView worked
RepairModelView worked

melissaquezada94 commented on core/data/src/main/java/com/iteneum/RepairModel.kt:
```diff
@@ -0,0 +1,20 @@
+ */
+
+data class RepairModel (
+    var unitDepartment: String = "",
```
Usually, the data class's attributes are of val type, we declare them var if it is really necessary.

JoseGR98 replied:
Okay, got it! Changed! :)

melissaquezada94 commented on feature/repair/src/main/java/com/iteneum/repair/data/RepairViewModel.kt:
```diff
@@ -0,0 +1,83 @@
+
+@HiltViewModel
+class RepairViewModel @Inject constructor(): ViewModel() {
+    private val _dataInfo : MutableState<RepairModel?> = mutableStateOf(null)
```
The repair model could ve null?

JoseGR98 replied:
For the moment it will be like that.
Based in Sergio & Carlos comments, he instructed to add the Unit to the RepairView, since it has to show the Department/House number. Because of that, it will be mostly probable that the View will have two type of perspectives. For the Employer, to be able to manipulate the field; and for the employee, to see the value already inserted. In case that it cannot be attached or found, we have to enable the null possibility.
What I will add for now is a TODO to verify this later on.

melissaquezada94 replied:
Ok so this case is only for the attribute unitDartment, right? So in this case why don't you send just the variable unitDepartment: String? instead all the data class

JoseGR98 replied:
Agree. I will make the change

JoseGR98 replied:
I made the change that instead of creating it null, I will make an initialization of the RepairModel data class. Is this better?

New declaration
![image](https://github.com/irv205/droid_leaspert/assets/25538206/2835ed97-35d0-4c84-b4d3-5a5b37c113fe)

JoseGR98 replied:
Change made but pending for new comments

melissaquezada94 commented on feature/repair/src/main/java/com/iteneum/repair/presentation/RepairView.kt:
```diff
@@ -8,34 +8,58 @@ import androidx.compose.ui.Modifier
 @Composable
-fun RepairView() {
+fun RepairContainer(
+    dataInfo: RepairModel?,
```
You are not using this variable: `dataInfo` to show the information when the view charged

JoseGR98 replied:
Yes, you are right. But I wrote on the TODO:
/* TODO - LpOutlinedTextFieldInput (Check component required) - To enable to action of passing value argument --> value = dataInfo?.unitDepartment ?: "000" */

There is a problem related to obtain the Unit (“Department number”) from the data source and paste it on the Repair View (as desired by Sergio). The component used to have the Unit does not have the actual “value” parameter. We need to add it so we can use it. This involves however the modification of the actual component.

The parameter value is not on the function parameters. It is required to be able to pass and use the dataInfo.
![image](https://user-images.githubusercontent.com/25538206/236571781-a0ebce86-27a5-4f6b-a4e2-8008d7a5964a.png)

melissaquezada94 replied:
You can put your task in block status because we are working on these changes to set the value to the input. 
Another thing is why are you using this parameter: ` dataInfo: RepairModel?,` because the only use that I see is to show the unit deparmet and if you are using only for this so I recommend you to change the name variable to `unitDepartment: String`

JoseGR98 replied:
I agree. Ok, will change to blocked and also make the change of unitDepartment.

melissaquezada94 commented on feature/repair/src/main/java/com/iteneum/repair/presentation/RepairView.kt:
```diff
@@ -63,17 +88,19 @@ fun RepairView() {
                     top = sizes.extraSize14
                 ),
-                value = contactPhoneNumberValue,
+                value = repairViewModel.repairModel.contactPhone,
```
You are not observing the contatPhone when it changes

JoseGR98 replied:
I did not get this point, sorry. I thougth that is showing correctly. Because i have it on two parameters.

value (So it can print the phoneNumber when inserting/deleting a number
![image](https://user-images.githubusercontent.com/25538206/236572244-4b94bffe-c0e0-48e9-b64b-637169517f2a.png)

onPhoneChanged (So it can be updated when the user inserts another value)
![image](https://user-images.githubusercontent.com/25538206/236572321-e37393a9-aac8-4f1a-aacf-8c8a38219a59.png)

melissaquezada94 replied:
Well, I mean, that you need to create an remember in this way: 
viewModel: 
 `private val _uiState = mutableStateOf(UiStateMainActivity(data = test, loading = true))
    val uiState: State<UiStateMainActivity> = _uiState`
    
    `fun updateLoading(){
        _uiState.value = _uiState.value.copy(
            loading = !_uiState.value.loading
        )
    }`
Compose: 
`private val viewModel by viewModels<MainActivityViewModel>()`
`val uiState by remember { viewModel.uiState }`
`Button(
                        onClick = { viewModel.updateLoading() },
                        modifier = Modifier.padding(bottom = 20.dp)
                    ) {
                        Text(text = "Click ${uiState.loading}")
                    }`

JoseGR98 replied:
Ok, will make the change. Thank you!

irv205 replied:
Status?

JoseGR98 replied:
Not sure of the meaning of this one. I would like to ask for assistance @irv205

JoseGR98 replied:
Modified. Done! (Help provided by Carlos)

melissaquezada94 commented on feature/repair/src/main/java/com/iteneum/repair/data/RepairViewModel.kt:
```diff
@@ -0,0 +1,83 @@
+        repairModel = repairModel.copy(category = category)
+    }
+    fun getProblemDescriptionFromView(problemDescription:String) {
+        repairModel = repairModel.copy(unitDepartment = problemDescription)
```
You are updating the wrong attibute plese check all

JoseGR98 replied:
That was a typo, my apologies. Corrected!

![image](https://user-images.githubusercontent.com/25538206/236572552-4e1042a5-5dad-4f82-b485-b771db55c75d.png)

JoseGR98 committed:
RepairModel and RepairViewModel corrections of code reviewer

JoseGR98 committed:
Merge branch 'develop' into feature/lp-114-repair-request-viewmodel

melissaquezada94 commented on feature/repair/src/main/java/com/iteneum/repair/data/RepairViewModel.kt:
```diff
@@ -0,0 +1,83 @@
+}
+
+val response: DataState<RepairModel> = DataState.Success(
+    RepairModel (
```
Why are you using a default RepairModel? My question is because when I navigate to your view I'm not able to see all this values in the corresponding inputs

JoseGR98 replied:
You are right, I only use the "unitDeparment" value, so will reduce it to that only.

irv205 replied:
What is the status?

JoseGR98 replied:
Done! Uploaded already

melissaquezada94 commented on feature/repair/src/main/java/com/iteneum/repair/presentation/RepairView.kt:
```diff
@@ -8,34 +8,58 @@ import androidx.compose.ui.Modifier
 @OptIn(ExperimentalMaterial3Api::class)
 @Composable
-fun RepairView() {
+fun RepairContainer(
```
In this function, I don't see any remember for each input value. We need a remember for each variable being observed from the viewModel to display the information that we get from the server( in this case is mocked) and also to be updated. This is because the information should be getting and updating from the viewModel and the View just need to show or call a updated method from the view and the information should be synchronized. 
Like in this case mostly of the components have it owns remember, I talk to Irving to change this so you can start working on add a remember of each variable that it is observer by the viewModel.

JoseGR98 replied:
Could you give me a structure example? Because since I am using compose I thought that the observers were this methods on file "RepairViewModel"

![image](https://github.com/irv205/droid_leaspert/assets/25538206/4c913300-9179-4ec0-ab3c-4bf1b053e0a2)


Or do you want something like this other example like when using Live Data for observables?
![image](https://github.com/irv205/droid_leaspert/assets/25538206/319f60ab-a4dd-4e1e-be9e-c300b4ebec32)

I don't know a lot of this yet, so your comments are really helpful. Thank you!

irv205 replied:
What is the status?

JoseGR98 replied:
Not sure about this one. I thought that _remember_ is not required anymore since the "ViewModel" is the one that handles that data. As I understood, the View does not require to remember any value. @irv205 
Therefore, I do not see the necessity of adding remember here.

JoseGR98 replied:
ViewModel reworked with events. Aided by Carlos. Done!

JoseGR98 committed:
Merge branch 'develop' into feature/lp-114-repair-request-viewmodel

JoseGR98 committed:
Merge branch 'develop' into feature/lp-114-repair-request-viewmodel

JoseGR98 committed:
Improvements on RepairView and RepairViewModel

irv205 commented on core/data/src/main/java/com/iteneum/RepairModel.kt:
```diff
@@ -0,0 +1,20 @@
+ * @author Jose Miguel Garcia Reyes
+ */
+
+data class RepairModel (
```
Rename to RepairRequest

JoseGR98 replied:
Ok, done!

irv205 commented on feature/repair/src/main/java/com/iteneum/repair/data/RepairViewModel.kt:
```diff
@@ -0,0 +1,81 @@
+
+@HiltViewModel
+class RepairViewModel @Inject constructor(): ViewModel() {
+    private val _repairModelMutable : MutableState<RepairModel> = mutableStateOf(RepairModel())
+    val repairModelRead: MutableState<RepairModel>
+        get() = _repairModelMutable
```
Rename to "repairRequest"

JoseGR98 replied:
Ok, done as well!

carlosph816 commented on feature/repair/src/main/java/com/iteneum/repair/data/RepairViewModel.kt:
```diff
@@ -0,0 +1,81 @@
+
+@HiltViewModel
+class RepairViewModel @Inject constructor(): ViewModel() {
+    private val _repairModelMutable : MutableState<RepairModel> = mutableStateOf(RepairModel())
```
if we are using a mutbleStateOf we cand change the implementation like

var  myData: RepairModel by mutableStateOf(RepairModel)
        private set

JoseGR98 replied:
Ok done! Thanks for the explanation in person as well

carlosph816 commented on feature/repair/src/main/java/com/iteneum/repair/presentation/RepairView.kt:
```diff
@@ -7,35 +7,57 @@ import androidx.compose.runtime.*
-fun RepairView() {
+fun RepairContainer(
+    unitDepartment: String,
+    repairViewModel: RepairViewModel = hiltViewModel(),
```
We don't need send the viewModel to a second composable view, viewModel always must be in the first one composable view, if you want to send a value or an object to the second one you need to access to that value and then add a new param to the second composable function.

JoseGR98 replied:
Ok, modified! Aided by you, thank you very much!

carlosph816 commented on feature/repair/src/main/java/com/iteneum/repair/presentation/RepairView.kt:
```diff
@@ -99,7 +129,9 @@ fun RepairView() {
                 items = optionsCategory.toList(),
-                selected = {/* TODO - Category field - To verify if extra functionality required */ }
+                selected = {
+                    repairViewModel.getCategoryFromView(it)
```
we need to use hight order function here and send the event to the top

JoseGR98 replied:
Modified! Thank you for your help

carlosph816 commented on feature/repair/src/main/java/com/iteneum/repair/presentation/RepairView.kt:
```diff
@@ -110,7 +142,9 @@ fun RepairView() {
                 hint = stringResource(id = R.string.hint_description),
-                onValueChanged = {/* TODO - Description field - To verify if extra functionality required */ }
+                onValueChanged = {
+                    repairViewModel.getProblemDescriptionFromView(it)
```
we need to use hight order function here and send the event to the top

carlosph816 commented on feature/repair/src/main/java/com/iteneum/repair/presentation/RepairView.kt:
```diff
@@ -126,7 +160,9 @@ fun RepairView() {
                 mimeTypes = arrayOf("video/*", "image/*"),
-                onFileSelected = {/* TODO - File Button - Add functionality to save files */ }
+                onFileSelected = {
+                    repairViewModel.getImageOrVideoFileFromView(it)
```
we need to use hight order function here and send the event to the top

JoseGR98 replied:
Modified! Thank you for your help

carlosph816 commented on feature/repair/src/main/java/com/iteneum/repair/presentation/RepairView.kt:
```diff
@@ -141,9 +177,9 @@ fun RepairView() {
+                selectedOption = repairViewModel.repairModel.permissionToEnter,
                 onOptionSelected = {
-                    optionSelectedRadioButtons = it
+                    repairViewModel.getPermissionToEnterFromView(it)
```
we need to use hight order function here and send the event to the top

JoseGR98 replied:
Modified! Thank you for your help

carlosph816 commented on feature/repair/src/main/java/com/iteneum/repair/presentation/RepairView.kt:
```diff
@@ -154,8 +190,11 @@ fun RepairView() {
-                onClicked = {/* TODO - Send Repair Button - Add functionality to save repair request info */ }
+                onClicked = {
+                    repairViewModel.onClickSendButton()
+                    navigateToApartment()
```
we need to use hight order function here and send the event to the top

JoseGR98 replied:
Modified! Thank you for your help

JoseGR98 committed:
Merge branch 'develop' into feature/lp-114-repair-request-viewmodel

JoseGR98 committed:
LP-109 Progress. Big and better ViewModel improvements.
- UIEventRepair.kt added
- RepairRequest.kt modified
- RepairView.kt modified
- RepairViewModel.kt modified (big changes)

JoseGR98 commented:
Waiting for Code Review, during next week. @irv205 
Waiting for Code Review at May 15th 2023

JoseGR98 committed:
Merge branch 'develop' into feature/lp-114-repair-request-viewmodel

JoseGR98 commented:
1 code reviewer approved. Waiting for one more or new feedback. I will need your approval @irv205

Lupillo17 commented on core/data/src/main/java/com/iteneum/UIEventRepair.kt:
```diff
@@ -0,0 +1,19 @@
+ * File that contains [UIEventRepair] sealed class
+ * It contains definitions of all the possible Events that will be catch on RepairView.
+ *
+ * @author Jose Miguel Garcia Reyes
```
Should be only 1 author, change it for modifiedBy

JoseGR98 replied:
Ok, removed!

Lupillo17 commented on feature/repair/src/main/java/com/iteneum/repair/data/RepairViewModel.kt:
```diff
@@ -0,0 +1,80 @@
+ * File that contains [RepairViewModel] that works as ViewModel for the Repair View.
+ * It will handle the data work & process when loading & be done with the view.
+ *
+ * @author Jose Miguel Garcia Reyes
```
Should be only 1 author, change it for modifiedBy

JoseGR98 replied:
Ok, updated :)

Lupillo17 commented on feature/repair/src/main/java/com/iteneum/repair/data/RepairViewModel.kt:
```diff
@@ -0,0 +1,80 @@
+        private set
+
+    fun getInformation() = viewModelScope.launch {
+        when (response) {
```
line 31, add remining states for response, acctually is just success state, you should add error state and loading state

JoseGR98 replied:
Ok, added!

Lupillo17 commented on feature/repair/src/main/java/com/iteneum/repair/presentation/RepairView.kt:
```diff
@@ -7,35 +7,68 @@ import androidx.compose.runtime.*
+ * @param navigateToApartment - to get the function to navigate ApartmentView when clicking button
+ * @param repairViewModel - to get the Repair View Model access
+ *
+ * Function [RepairContainer] has
```
Move RepairContainer documentation to their respective function.

JoseGR98 replied:
Ok, moved bellow

Lupillo17 commented on feature/repair/src/main/java/com/iteneum/repair/presentation/RepairView.kt:
```diff
@@ -125,8 +169,10 @@ fun RepairView() {
                     .fillMaxWidth(),
-                mimeTypes = arrayOf("video/*", "image/*"),
-                onFileSelected = {/* TODO - File Button - Add functionality to save files */ }
+                mimeTypes = arrayOf("video/*", "image/*"), /* TODO - Verify how will this work */
```
move arrayOf to the viewModel and retrive it as value.

JoseGR98 replied:
Ok, done!

Lupillo17 commented on core/data/src/main/java/com/iteneum/RepairRequest.kt:
By the moment move this file to **feature/repair** package, not in **core package**

JoseGR98 replied:
Ok, moved!

Lupillo17 commented on core/data/src/main/java/com/iteneum/UIEventRepair.kt:
By the moment move this file to **feature/repair** package, not in **core package**

JoseGR98 replied:
Ok, moved!

Lupillo17 commented on feature/repair/src/main/java/com/iteneum/repair/data/RepairViewModel.kt:
```diff
@@ -0,0 +1,80 @@
+}
+
+val response: DataState<RepairRequest> = DataState.Success(
+    RepairRequest(
```
add missing parameters to **RepairRequest** object

JoseGR98 replied:
Ok. They are not required because user will fill them, but added.

Lupillo17 commented on feature/repair/src/main/java/com/iteneum/repair/presentation/RepairView.kt:
```diff
@@ -7,35 +7,68 @@ import androidx.compose.runtime.*
-    var contactPhoneNumberValue by remember {
-        mutableStateOf("")
-    }
     LeasePertTheme {
```
I think **LeasePertTheme** is not necesary, try delete it and it should works well

JoseGR98 replied:
Ok, removed it!

Lupillo17 commented on feature/repair/src/main/java/com/iteneum/repair/presentation/RepairView.kt:
```diff
@@ -7,35 +7,68 @@ import androidx.compose.runtime.*
-        mutableStateOf("")
-    }
     LeasePertTheme {
         Column(
```
I can't scroll the view, please add vertical scrolling to **Column Modifier**

JoseGR98 replied:
Ok, added!

JoseGR98 committed:
Merge branch 'develop' into feature/lp-114-repair-request-viewmodel

JoseGR98 committed:
LP-114 Code Review updates made by Lupillo17

JoseGR98 commented:
Changed updated! Waiting for new code approbal @Lupillo17

Lupillo17 commented on feature/repair/src/main/java/com/iteneum/repair/data/RepairViewModel.kt:
```diff
@@ -19,19 +20,27 @@ import javax.inject.Inject
                 repairRequest = response.data
             }
+            is DataState.Error -> {
+                repairRequest = RepairRequest(unitDepartment = "EEE")
```
By the moment add a TODO and delete line 39

JoseGR98 replied:
Ok, done!

Lupillo17 commented on feature/repair/src/main/java/com/iteneum/repair/data/RepairViewModel.kt:
```diff
@@ -19,19 +20,27 @@ import javax.inject.Inject
+                repairRequest = RepairRequest(unitDepartment = "EEE")
+            }
+            is DataState.Loading -> {
+                repairRequest = RepairRequest(unitDepartment = "LLL")
```
By the moment add a TODO and delete line 42

JoseGR98 replied:
Ok,done!

Lupillo17 commented on feature/repair/src/main/java/com/iteneum/repair/presentation/RepairView.kt:
```diff
@@ -47,17 +42,30 @@ fun RepairView(
+ * Function [RepairContainer] has
+ * @param formularyValues - to pass the content of the RepairRequest
+ * @param formularyData - to handle the UI events of data when text is modified
+ * @param validatePhone - to validate is thw phone contains 10 digits and only numbers
```
check orthography **thw phone**

JoseGR98 replied:
Ok, corrected!

Lupillo17 commented on feature/repair/src/main/java/com/iteneum/repair/presentation/RepairView.kt:
```diff
@@ -47,17 +42,30 @@ fun RepairView(
 }
 
+/**
+ * Function [RepairContainer] has
```
Describe the main pourpose of RepairContainer, for example: contains the view of repair screen... describe it

JoseGR98 replied:
Done!

JoseGR98 committed:
Merge branch 'develop' into feature/lp-114-repair-request-viewmodel

JoseGR98 commented:
New PR uploaded , addressed @Lupillo17

JoseGR98 committed:
LP-114 Code Review updates made by Lupillo17

Lupillo17 commented on feature/repair/src/main/java/com/iteneum/repair/presentation/RepairView.kt:
```diff
@@ -50,10 +50,11 @@ fun RepairView(
 
 /**
- * Function [RepairContainer] has
+ * Function [RepairContainer] contains the view os RepairView, will generate the composable elements
```
check orthography

JoseGR98 replied:
Checked and improved!

Lupillo17 commented on feature/repair/src/main/java/com/iteneum/repair/presentation/RepairView.kt:
```diff
@@ -50,10 +50,11 @@ fun RepairView(
  * @param formularyValues - to pass the content of the RepairRequest
  * @param formularyData - to handle the UI events of data when text is modified
- * @param validatePhone - to validate is thw phone contains 10 digits and only numbers
+ * @param validatePhone - to validate that value phone contains at least 10 digits and numbers only
```
check description and correct it please

JoseGR98 replied:
Corrected!

JoseGR98 committed:
Last modifications based on Code Review by Lupillo17
